In [3]:
# Installation for GPU llama-cpp-python
# uncomment and run the following code in case GPU is being used
!CMAKE_ARGS="-DLLAMA_CUBLAS=on" FORCE_CMAKE=1
!pip install llama-cpp-python

# Installation for CPU llama-cpp-python
# uncomment and run the following code in case GPU is not being used
# !CMAKE_ARGS="-DLLAMA_CUBLAS=off" FORCE_CMAKE=1 pip install llama-cpp-python==0.1.85 --force-reinstall --no-cache-dir -q


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.7/50.7 MB 41.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 3.8 MB/s eta 0:00:00
  Created wheel for llama-cpp-python: filename=llama_cpp_python-0.3.16-cp312-cp312-linux_x86_64.whl size=4500213 sha256=33a8cf828e64ed8ae997588f8add8b931ed0d680500a311fb2338fd7e4ddc5a3
  Stored in directory: /root/.cache/pip/wheels/90/82/ab/8784ee3fb99ddb07fd36a679ddbe63122cc07718f6c1eb3be8
Successfully built llama-cpp-python


In [4]:
# For installing the libraries & downloading models from HF Hub
!pip cache purge -q
!pip install huggingface_hub
!pip install tiktoken
!pip install pymupdf
!pip install langchain
!pip install langchain-community
!pip install chromadb
!pip install sentence-transformers
!pip install numpy
!pip install pandas


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.1/24.1 MB 121.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 35.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 45.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 471.2/471.2 kB 38.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.7/64.7 kB 6.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 4.9 MB/s eta 0:00:00
  Attempting uninstall: requests
    Found existing installation: requests 2.32.4
    Uninstalling requests-2.32.4:
      Successfully uninstalled requests-2.32.4
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.3.79
    Uninstalling langchain-core-0.3.79:
      Successfully uninstalled langchain-core-0.3.79
  Attempting uninstall: langchain-text-splitters
    Found existing installation: langchain-text-splitters 0.3.11
    Uninstalling langchain-text-splitters-0.3.11:
      Successfu

In [5]:
#Libraries for downloading and loading the llm
from huggingface_hub import hf_hub_download
from llama_cpp import Llama

In [6]:
#Downloading and Loading the model
model_name_or_path = "TheBloke/Mistral-7B-Instruct-v0.2-GGUF"
model_basename = "mistral-7b-instruct-v0.2.Q6_K.gguf"


model_path = hf_hub_download(
    repo_id=model_name_or_path,
    filename=model_basename
)

print(f"Model downloaded to: {model_path}")

llm = Llama(
    model_path=model_path,
    n_ctx=2300,
    n_gpu_layers=8,
    n_batch=128
)

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


mistral-7b-instruct-v0.2.Q6_K.gguf:   0%|          | 0.00/5.94G [00:00<?, ?B/s]

llama_model_loader: loaded meta data with 24 key-value pairs and 291 tensors from /root/.cache/huggingface/hub/models--TheBloke--Mistral-7B-Instruct-v0.2-GGUF/snapshots/3a6fbf4a41a1d52e415a4958cde6856d34b2db93/mistral-7b-instruct-v0.2.Q6_K.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.name str              = mistralai_mistral-7b-instruct-v0.2
llama_model_loader: - kv   2:                       llama.context_length u32              = 32768
llama_model_loader: - kv   3:                     llama.embedding_length u32              = 4096
llama_model_loader: - kv   4:                          llama.block_count u32              = 32
llama_model_loader: - kv   5:                  llama.feed_forward_length u32              = 14336
llama_model_loade

Model downloaded to: /root/.cache/huggingface/hub/models--TheBloke--Mistral-7B-Instruct-v0.2-GGUF/snapshots/3a6fbf4a41a1d52e415a4958cde6856d34b2db93/mistral-7b-instruct-v0.2.Q6_K.gguf


load_tensors:   CPU_Mapped model buffer size =  5666.09 MiB
...................................................................................................
llama_context: constructing llama_context
llama_context: n_seq_max     = 1
llama_context: n_ctx         = 2300
llama_context: n_ctx_per_seq = 2300
llama_context: n_batch       = 128
llama_context: n_ubatch      = 128
llama_context: causal_attn   = 1
llama_context: flash_attn    = 0
llama_context: kv_unified    = false
llama_context: freq_base     = 1000000.0
llama_context: freq_scale    = 1
llama_context: n_ctx_per_seq (2300) < n_ctx_train (32768) -- the full capacity of the model will not be utilized
set_abort_callback: call
llama_context:        CPU  output buffer size =     0.12 MiB
create_memory: n_ctx = 2304 (padded)
llama_kv_cache_unified: layer   0: dev = CPU
llama_kv_cache_unified: layer   1: dev = CPU
llama_kv_cache_unified: layer   2: dev = CPU
llama_kv_cache_unified: layer   3: dev = CPU
llama_kv_cache_unified: layer 

In [7]:
#Response
# Temparature is kept at 0 because for this business use case we want factual answers

def response(query,max_tokens=200,temperature=0,top_p=0.95,top_k=50):
    model_output = llm(
      prompt=query,
      max_tokens=max_tokens,
      temperature=temperature,
      top_p=top_p,
      top_k=top_k
    )

    return model_output['choices'][0]['text']

In [8]:
#Question 1: What is the protocol for managing sepsis in a critical care unit?
query1 = "What is the protocol for managing sepsis in a critical care unit?"
print(response(query1))

llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    1191.97 ms /    16 tokens (   74.50 ms per token,    13.42 tokens per second)
llama_perf_context_print:        eval time =   30341.09 ms /   199 runs   (  152.47 ms per token,     6.56 tokens per second)
llama_perf_context_print:       total time =   31655.48 ms /   215 tokens
llama_perf_context_print:    graphs reused =        192




Sepsis is a life-threatening condition that can arise from an infection, and it requires prompt recognition and aggressive management in a critical care unit. The following are the general steps for managing sepsis in a critical care unit:

1. Early recognition and suspicion: Septic patients may present with non-specific symptoms such as fever, chills, tachycardia, tachypnea, altered mental status, or lactic acidosis. It is essential to have a high index of suspicion for sepsis, especially in patients with known infections or risk factors.
2. Initial assessment and resuscitation: The first step in managing sepsis is to assess and resuscitate the patient. This includes assessing airway, breathing, circulation, and disability (ABCD) and providing appropriate interventions such as oxygen therapy, fluid resuscitation, and vasopressor support if necessary.
3. Source control:


In [9]:
#Question2: What are the common symptoms of appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?
query2 = "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(response(query2))

Llama.generate: 2 prefix-match hit, remaining 32 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    3488.81 ms /    32 tokens (  109.03 ms per token,     9.17 tokens per second)
llama_perf_context_print:        eval time =   30322.82 ms /   199 runs   (  152.38 ms per token,     6.56 tokens per second)
llama_perf_context_print:       total time =   33934.02 ms /   231 tokens
llama_perf_context_print:    graphs reused =        192




Appendicitis is a medical condition characterized by inflammation of the appendix, a small pouch that extends from the cecum, the first part of the large intestine. The symptoms of appendicitis can vary from person to person, but the following are the most common:

1. Abdominal pain: The pain is typically located in the lower right side of the abdomen and may be dull at first, but it can quickly become sharp and severe. The pain may worsen when you move, cough, or sneeze.
2. Loss of appetite: You may lose your appetite due to the abdominal pain or feel nauseous.
3. Nausea and vomiting: You may feel sick to your stomach and vomit.
4. Fever: A fever of 100.4°F (38°C) or higher is common with appendicitis.
5. Const


In [10]:
#Question3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?
query3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(response(query3))

Llama.generate: 4 prefix-match hit, remaining 34 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    2428.08 ms /    34 tokens (   71.41 ms per token,    14.00 tokens per second)
llama_perf_context_print:        eval time =   30503.12 ms /   199 runs   (  153.28 ms per token,     6.52 tokens per second)
llama_perf_context_print:       total time =   33053.27 ms /   233 tokens
llama_perf_context_print:    graphs reused =        192




Sudden patchy hair loss, also known as alopecia areata, is a common autoimmune disorder that affects the hair follicles, leading to hair loss in small, round patches on the scalp, beard, or other areas of the body. The exact cause of alopecia areata is not known, but it is believed to be related to a problem with the immune system.

There are several treatments that have been shown to be effective in addressing sudden patchy hair loss:

1. Corticosteroids: Corticosteroids are anti-inflammatory medications that can help reduce inflammation and suppress the immune system, allowing the hair follicles to regrow. They can be applied topically or taken orally.
2. Minoxidil: Minoxidil is a medication that has been shown to promote hair growth in some people with alopecia areata. It is applied


In [11]:
#Question4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?
query4 = "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(response(query4))

Llama.generate: 2 prefix-match hit, remaining 28 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    2021.32 ms /    28 tokens (   72.19 ms per token,    13.85 tokens per second)
llama_perf_context_print:        eval time =   30384.68 ms /   199 runs   (  152.69 ms per token,     6.55 tokens per second)
llama_perf_context_print:       total time =   32531.14 ms /   227 tokens
llama_perf_context_print:    graphs reused =        191




A person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function, is typically diagnosed with a traumatic brain injury (TBI). The treatment for a TBI depends on the severity and location of the injury, as well as the individual's overall health and age.

Immediate treatment for a TBI may include:

1. Emergency medical care: This may include surgery to remove hematomas or other obstructions, as well as treatment for other injuries.
2. Medications: Depending on the symptoms, medications may be prescribed to manage pain, reduce swelling, prevent or treat infections, or control seizures.
3. Rehabilitation: Rehabilitation may include physical therapy, occupational therapy, speech therapy, and cognitive rehabilitation to help the person regain lost skills and functions.
4. Supportive care: This may include assistance with daily activities


In [12]:
#Question5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?
query5 = "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(response(query5))

Llama.generate: 2 prefix-match hit, remaining 35 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    3667.78 ms /    35 tokens (  104.79 ms per token,     9.54 tokens per second)
llama_perf_context_print:        eval time =   30354.32 ms /   199 runs   (  152.53 ms per token,     6.56 tokens per second)
llama_perf_context_print:       total time =   34143.06 ms /   234 tokens
llama_perf_context_print:    graphs reused =        192




First and foremost, if a person has fractured their leg during a hiking trip, it is essential to ensure their safety and prevent further injury. Here are some necessary precautions and treatment steps:

1. Assess the situation: Check the extent of the injury and assess the person's condition. If the fracture is open or the person is in severe pain, immobilize the leg with a splint or a makeshift sling to prevent any movement.
2. Call for help: If possible, call for emergency medical assistance. If there is no cell phone reception, try to signal for help using a mirror, whistle, or other means.
3. Provide first aid: Apply a sterile dressing to the injury to prevent infection. If the fracture is open, apply pressure to stop any bleeding.
4. Immobilize the leg: Use a splint, a makeshift sling, or a tour


Observations:
As contextual data is not provided as input, LLama model is trying to give generalized answers.
The answers are medically appropriate and lists effective treatements.

However, the model tends to provide useful but somewhat high-level responses. This could make the answers seem more like generic guidelines rather than tailored advice for specific scenarios.

We will leverage prompt engineering to enhance the quality of our responses.

In [13]:
#Question Answering using LLM with Prompt Engineering
# Add instructions to the prompt for better response generation
# Given the sensitivity of medical diagnosis, I am adding a prompt to minimize hallucination
system_prompt = "You are a helpful and knowledgeable medical assistant. Answer the following medical question accurately and concisely based on common medical knowledge. If you don't know the answer, please state that you don't have enough information."
print ("System Prompt:" + system_prompt)

System Prompt:You are a helpful and knowledgeable medical assistant. Answer the following medical question accurately and concisely based on common medical knowledge. If you don't know the answer, please state that you don't have enough information.


In [14]:
#Question 1: What is the protocol for managing sepsis in a critical care unit?
query1 = system_prompt + "\n" + "What is the protocol for managing sepsis in a critical care unit?"
print(response(query1))

Llama.generate: 1 prefix-match hit, remaining 61 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    4275.86 ms /    61 tokens (   70.10 ms per token,    14.27 tokens per second)
llama_perf_context_print:        eval time =   30555.50 ms /   199 runs   (  153.55 ms per token,     6.51 tokens per second)
llama_perf_context_print:       total time =   34954.63 ms /   260 tokens
llama_perf_context_print:    graphs reused =        191




Sepsis is a life-threatening condition caused by a severe infection. In a critical care unit, managing sepsis involves the following steps:

1. Early recognition and diagnosis: Monitor vital signs, laboratory values, and clinical symptoms closely. Suspect sepsis in any patient with suspected or confirmed infection and organ dysfunction.
2. Immediate fluid resuscitation: Administer intravenous fluids to maintain adequate blood pressure and organ perfusion. The goal is to achieve a mean arterial pressure (MAP) of 65 mmHg or higher.
3. Antibiotics: Administer broad-spectrum antibiotics as soon as possible based on the suspected infection site and microbiological culture results.
4. Source control: Identify and address the source of infection, such as removing an infected catheter or draining an abscess.
5. Vasopressors: If fluid res


In [15]:
#Question2: What are the common symptoms of appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?
query2 = system_prompt + "\n" + "What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?"
print(response(query2))

Llama.generate: 48 prefix-match hit, remaining 32 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    2275.14 ms /    32 tokens (   71.10 ms per token,    14.07 tokens per second)
llama_perf_context_print:        eval time =   30523.95 ms /   199 runs   (  153.39 ms per token,     6.52 tokens per second)
llama_perf_context_print:       total time =   32919.12 ms /   231 tokens
llama_perf_context_print:    graphs reused =        192




Appendicitis is a common inflammatory condition of the appendix, a small pouch located in the lower right side of the abdomen. The symptoms of appendicitis can include:

1. Sudden pain in the lower right abdomen, which may start as a mild ache and gradually develop into a sharp pain.
2. Loss of appetite and feeling sick to your stomach (nausea).
3. Fever, which may be low-grade at first but can rise as high as 103°F (39.4°C).
4. Vomiting.
5. Constipation or diarrhea.
6. Abdominal swelling and tenderness.
7. Inability to pass gas or have a bowel movement.
8. Pain in the lower back, on the right side.
9. Feeling restless or unable to find a comfortable position.


In [16]:
#Question3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?
query3 = system_prompt + "\n" + "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(response(query3))

Llama.generate: 50 prefix-match hit, remaining 34 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    2456.14 ms /    34 tokens (   72.24 ms per token,    13.84 tokens per second)
llama_perf_context_print:        eval time =   30460.14 ms /   199 runs   (  153.07 ms per token,     6.53 tokens per second)
llama_perf_context_print:       total time =   33038.28 ms /   233 tokens
llama_perf_context_print:    graphs reused =        192




Sudden patchy hair loss, also known as alopecia areata, is an autoimmune condition that causes hair loss in small, round patches on the scalp, beard, or other areas of the body. The exact cause is unknown, but it's believed to be related to a problem with the immune system.

Effective treatments for addressing sudden patchy hair loss include:

1. Corticosteroids: These are anti-inflammatory medications that can help reduce inflammation and suppress the immune system response. They can be applied topically or taken orally.
2. Immunotherapy: This involves the use of medications that stimulate the immune system to attack the hair loss. One such medication is minoxidil.
3. Hair transplantation: This is a surgical procedure in which healthy hair is transplanted from one area of the scalp to another. It's usually considered a


In [17]:
#Question4: What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?

query4 = system_prompt + "\n" + "What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?"
print(response(query4))

Llama.generate: 48 prefix-match hit, remaining 28 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    2118.06 ms /    28 tokens (   75.65 ms per token,    13.22 tokens per second)
llama_perf_context_print:        eval time =   30626.58 ms /   199 runs   (  153.90 ms per token,     6.50 tokens per second)
llama_perf_context_print:       total time =   32868.46 ms /   227 tokens
llama_perf_context_print:    graphs reused =        192




For a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function, the recommended treatments depend on the specific type and severity of the injury. Here are some common treatments:

1. Emergency care: For severe brain injuries, the first priority is to provide emergency care to stabilize the patient's vital signs and prevent further damage. This may include surgery to remove hematomas or other obstructions, administering oxygen, and controlling seizures.
2. Medications: Depending on the symptoms, medications may be prescribed to manage various conditions such as pain, seizures, infections, and depression.
3. Rehabilitation: Rehabilitation is an essential part of treatment for brain injuries. It may include physical therapy to help the patient regain mobility, occupational therapy to help with daily living activities, speech therapy to improve communication skills, and cognitive therapy to help with memory and problem-

In [18]:
#Question5: What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?

query5 = system_prompt + "\n" + "What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?"
print(response(query5))

Llama.generate: 48 prefix-match hit, remaining 35 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =    2512.34 ms /    35 tokens (   71.78 ms per token,    13.93 tokens per second)
llama_perf_context_print:        eval time =   30600.41 ms /   199 runs   (  153.77 ms per token,     6.50 tokens per second)
llama_perf_context_print:       total time =   33234.86 ms /   234 tokens
llama_perf_context_print:    graphs reused =        192




A leg fracture during a hiking trip requires prompt medical attention. Here are the necessary precautions and treatment steps:

1. Immobilize the fracture: Use a splint, sling, or a brace to prevent any movement of the affected leg. This will help reduce pain, prevent further damage, and promote proper healing.

2. Control bleeding: Apply direct pressure to the wound with a clean cloth to control any bleeding. Elevate the injured leg above heart level to help reduce swelling and bleeding.

3. Seek medical help: If the fracture is severe or if there are signs of shock (pale skin, rapid heartbeat, rapid breathing, or loss of consciousness), call for emergency medical assistance immediately.

4. Pain management: Use over-the-counter pain relievers, such as acetaminophen or ibuprofen, to help manage pain. Your healthcare provider may also prescribe


Observations:
1.With system prompt we can see model is giving back the matching responses from different sources of data.
2. Overall, answers are superior with system prompt in terms of completeness, clinical clarity, and usefulness for decision-making in a critical care setting.

In [19]:
import sys
sys.setrecursionlimit(2000)

## Data Preparation for RAG
### Loading the Data
#Libraries for processing dataframes, text
import json, os
import tiktoken
import pandas as pd

!pip install langchain

#Libraries for Loading Data, Chunking, Embedding, and Vector Databases
from langchain_text_splitters import RecursiveCharacterTextSplitter
from langchain_community.document_loaders import PyMuPDFLoader
from langchain_community.embeddings.sentence_transformer import SentenceTransformerEmbeddings
from langchain_community.vectorstores import Chroma

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 449.8/449.8 kB 9.8 MB/s eta 0:00:00
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 1.0.4
    Uninstalling langchain-core-1.0.4:
      Successfully uninstalled langchain-core-1.0.4
  Attempting uninstall: langchain-text-splitters
    Found existing installation: langchain-text-splitters 1.0.0
    Uninstalling langchain-text-splitters-1.0.0:
      Successfully uninstalled langchain-text-splitters-1.0.0
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langchain-classic 1.0.0 requires langchain-core<2.0.0,>=1.0.0, but you have langchain-core 0.3.79 which is incompatible.
langchain-classic 1.0.0 requires langchain-text-splitters<2.0.0,>=1.0.0, but you have langchain-text-splitters 0.3.11 which is incompatible.
langchain-community 0.4.1 requires langchain-core<2.0.0,>=1.0.1, but you ha

In [20]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [21]:
## Data Overview


pdf_path = '/content/drive/My Drive/Colab Notebooks/medical_diagnosis_manual.pdf'
if not os.path.exists(pdf_path):
    raise FileNotFoundError(f"PDF not found at {pdf_path}")

# Use the PyMuPDFLoader to load the document
try:
    # Note: Loading large PDFs (>4,000 pages) may require significant memory. Consider chunked processing if RAM is limited.
    loader = PyMuPDFLoader(pdf_path)
    documents = loader.load()
    print(f"Loaded {len(documents)} pages from the PDF.")
except Exception as e:
    print(f"Error loading PDF: {e}")
    raise

Loaded 4114 pages from the PDF.


In [22]:
# Check first 5 pages (or fewer if PDF is smaller) for debugging
for i in range(min(5, len(documents))):
    print(f"--- Page {i+1} ---")
    print(documents[i].page_content[:500] + "...")
print(f"Number of pages: {len(documents)}")

--- Page 1 ---
divyapathipati@gmail.com
ILZ0KBS6PR
ant for personal use by divyapathipati@g
shing the contents in part or full is liable...
--- Page 2 ---
divyapathipati@gmail.com
ILZ0KBS6PR
This file is meant for personal use by divyapathipati@gmail.com only.
Sharing or publishing the contents in part or full is liable for legal action....
--- Page 3 ---
Table of Contents
1
Front    ................................................................................................................................................................................................................
1
Cover    .......................................................................................................................................................................................................
2
Front Matter    ....................................
--- Page 4 ---
491
Chapter 44. Foot & Ankle Disorders    .................................................................................

In [23]:
## Data chunking

def get_data_chunks(
    data,
    chunk_size=1000,
    chunk_overlap=200,
    split_method="recursive",
    separators=["\n\n", "\n", ". ", " ", ""],
    min_chunk_size=50,
    respect_sentence_boundaries=True,
    respect_paragraph_boundaries=True,
    length_function=len,
    max_chunks=None,
    add_metadata=True
):
    """
    Splits a list of documents into smaller chunks using a specified method.

    Args:
        data: A list of document objects (e.g., from Langchain loaders).
        chunk_size: The maximum size of each chunk.
        chunk_overlap: The number of characters to overlap between chunks.
        split_method: The method to use for splitting ('recursive').
        separators: A list of separators to use for splitting.
        min_chunk_size: The minimum size of each chunk.
        respect_sentence_boundaries: Whether to try to split on sentence boundaries.
        respect_paragraph_boundaries: Whether to try to split on paragraph boundaries.
        length_function: The function to use to measure chunk length.
        max_chunks: The maximum number of chunks to generate.
        add_metadata: Whether to add metadata to the chunks.

    Returns:
        A list of chunked documents.
    """
    try:
        if split_method == "recursive":
            text_splitter = RecursiveCharacterTextSplitter(
                chunk_size=chunk_size,
                chunk_overlap=chunk_overlap,
                length_function=length_function,
                is_separator_regex=False,
                separators=separators
            )
            chunks = text_splitter.split_documents(data)
        else:
            raise ValueError(f"Unsupported split_method: {split_method}")
    except Exception as e:
        print(f"Error during chunking: {e}")
        raise
    chunks = [chunk for chunk in chunks if length_function(chunk.page_content) >= min_chunk_size]
    if add_metadata:
        for chunk in chunks:
            if not chunk.metadata:
                chunk.metadata = {"source": "medical_diagnosis_manual.pdf", "page": 0}  # Fallback
    if max_chunks is not None:
        chunks = chunks[:max_chunks]
    return chunks

In [24]:
# create data chunks for loaded pdf
chunk_size = 500  # chunk size
chunk_overlap = 100 # chunk overlap

chunks = get_data_chunks(
    documents,
    chunk_size=chunk_size,
    chunk_overlap=chunk_overlap,
    split_method="recursive",
    separators=["\n\n", "\n", ". ", " ", ""],
    min_chunk_size=50,
    respect_sentence_boundaries=True,
    respect_paragraph_boundaries=True,
    length_function=len,
    max_chunks=None, # Process all chunks
    add_metadata=True
)

# Print some validation and check statements
print(f"\n--- Chunking Validation ---")
print(f"Number of documents loaded: {len(documents)}")
print(f"Number of chunks created: {len(chunks)}")

# Check the first few chunks
print(f"\nFirst {min(5, len(chunks))} chunks:")
for i in range(min(5, len(chunks))):
  print(f"--- Chunk {i+1} ---")
  print(f"Chunk length: {len(chunks[i].page_content)}")
  print(f"Chunk metadata: {chunks[i].metadata}")
  print(chunks[i].page_content[:200] + "...") # Print first 200 characters of the chunk content

# Check the last few chunks (if there are more than 5)
if len(chunks) > 5:
    print(f"\nLast {min(5, len(chunks)-5)} chunks:")
    for i in range(max(0, len(chunks)-5), len(chunks)):
        print(f"--- Chunk {i+1} ---")
        print(f"Chunk length: {len(chunks[i].page_content)}")
        print(f"Chunk metadata: {chunks[i].metadata}")
        print(chunks[i].page_content[:200] + "...") # Print first 200 characters of the chunk content

# Additional checks
if len(chunks) > 0:
    # Check minimum chunk size
    min_len = min(len(chunk.page_content) for chunk in chunks)
    print(f"\nMinimum chunk length: {min_len}")
    if min_len < 50: # Based on min_chunk_size parameter
        print("Warning: Some chunks might be smaller than the specified minimum size.")

    # Check for empty chunks
    empty_chunks = sum(1 for chunk in chunks if len(chunk.page_content) == 0)
    print(f"Number of empty chunks: {empty_chunks}")

    # Check metadata presence (assuming add_metadata is True)
    metadata_missing = sum(1 for chunk in chunks if not hasattr(chunk, 'metadata') or not chunk.metadata)
    print(f"Number of chunks missing metadata: {metadata_missing}")
else:
    print("\nNo chunks were created.")


--- Chunking Validation ---
Number of documents loaded: 4114
Number of chunks created: 34532

First 5 chunks:
--- Chunk 1 ---
Chunk length: 121
Chunk metadata: {'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'creator': 'Atop CHM to PDF Converter', 'creationdate': '2012-06-15T05:44:40+00:00', 'source': '/content/drive/My Drive/Colab Notebooks/medical_diagnosis_manual.pdf', 'file_path': '/content/drive/My Drive/Colab Notebooks/medical_diagnosis_manual.pdf', 'total_pages': 4114, 'format': 'PDF 1.7', 'title': 'The Merck Manual of Diagnosis & Therapy, 19th Edition', 'author': '', 'subject': '', 'keywords': '', 'moddate': '2025-11-08T23:58:26+00:00', 'trapped': '', 'modDate': 'D:20251108235826Z', 'creationDate': 'D:20120615054440Z', 'page': 0}
divyapathipati@gmail.com
ILZ0KBS6PR
ant for personal use by divyapathipati@g
shing the contents in part or full is liable...
--- Chunk 2 ---
Chunk length: 184
Chunk metadata: {'producer': 'pdf-lib (https://github.com/Hopding/pdf-lib)', 'c

In [25]:
#EMBEDDING
# Note: For M4 with 16GB RAM, process chunks in small batches to avoid memory issues.
embedding_function = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")

# Embed chunks in batches
embedded_chunks = []
batch_size = 100  # Safe for 16GB RAM
for i in range(0, len(chunks), batch_size):
    batch = chunks[i:i+batch_size]
    for j, chunk in enumerate(batch):
        try:
            embedding = embedding_function.embed_query(chunk.page_content)
            embedded_chunks.append((chunk, embedding))
        except Exception as e:
            print(f"Error embedding chunk {i+j}: {e}")
print(f"\n--- Embedding Validation ---")
print(f"Number of original chunks: {len(chunks)}")
print(f"Number of successfully embedded chunks: {len(embedded_chunks)}")
if len(embedded_chunks) == len(chunks):
    print("All chunks successfully embedded.")
else:
    print(f"Embedded {len(embedded_chunks)}/{len(chunks)} chunks.")

if len(embedded_chunks) > 0:
    first_embedded_chunk, first_embedding = embedded_chunks[0]
    print(f"Type of first embedding: {type(first_embedding)}")
    import numpy as np
    first_embedding_np = np.array(first_embedding)
    print(f"Dimension of first embedding: {len(first_embedding_np)}")
    print(f"First 10 values: {first_embedding_np[:10]}")

/tmp/ipython-input-255294834.py:3: LangChainDeprecationWarning: The class `HuggingFaceEmbeddings` was deprecated in LangChain 0.2.2 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-huggingface package and should be used instead. To use it run `pip install -U :class:`~langchain-huggingface` and import as `from :class:`~langchain_huggingface import HuggingFaceEmbeddings``.
  embedding_function = SentenceTransformerEmbeddings(model_name="all-MiniLM-L6-v2")


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]


--- Embedding Validation ---
Number of original chunks: 34532
Number of successfully embedded chunks: 34532
All chunks successfully embedded.
Type of first embedding: <class 'list'>
Dimension of first embedding: 384
First 10 values: [-0.08746312  0.08473203  0.06812345 -0.03073913  0.00991669 -0.10547927
  0.08597927  0.12860115 -0.03966499  0.0078195 ]


Observations:
No. of chunks created are 34532 and all chunks are successfully embedded.

In [26]:
#Next step is loading vectorized chunks in Vector Database
import os, shutil

from langchain_community.vectorstores import Chroma

persist_directory = 'medical_db'
if os.path.exists(persist_directory):
    print(f"Removing existing database at {persist_directory}")
    shutil.rmtree(persist_directory, ignore_errors=True)

if not os.path.exists(persist_directory):
    os.makedirs(persist_directory)
    print(f"Created directory at {persist_directory}")

try:
    vector_db = Chroma.from_documents(
        documents=[chunk for chunk, embedding in embedded_chunks],
        embedding=embedding_function,
        persist_directory=persist_directory
    )
except Exception as e:
    print(f"Error creating Chroma database: {e}")
    raise

# Simplified vector database validation
print(f"\n--- Vector Database Validation ---")
if os.path.exists(persist_directory):
    print(f"Chroma database created at {persist_directory}.")
try:
    count = vector_db._collection.count()
    print(f"Number of items in database: {count}")
    print("Match with embedded chunks." if count == len(embedded_chunks) else "Item count mismatch.")
except Exception as e:
    print(f"Error retrieving database count: {e}")

Created directory at medical_db

--- Vector Database Validation ---
Chroma database created at medical_db.
Number of items in database: 34532
Match with embedded chunks.


Vector database has all chunks stored - 34532.

In [27]:
#Next step is retriever
# To retrieve vectorized response from vector db (retrieve top 3 documents)
retriever = vector_db.as_retriever(
    search_type="similarity",  # Using similarity search
    search_kwargs={"k": 3}     # Retrieve top 3 similar documents
)

# --- Validation and Conclusions ---
print(f"\n--- Retriever Validation ---")
if retriever:
    print("Conclusion: Retriever successfully created.")
    try:
        # Test with queries from problem statement (e.g., sepsis, appendicitis, hair loss)
        sample_query = "What are the symptoms of appendicitis?"
        retrieved_docs = retriever.invoke(sample_query)
        print(f"\nRetrieved {len(retrieved_docs)} documents for a sample query.")
        if len(retrieved_docs) > 0:
            print("Sample of retrieved document content (first 100 chars):")
            print(retrieved_docs[0].page_content[:100] + "...")
        else:
            print("No documents retrieved. Check vector database content or query relevance.")

        print(f"\nRetriever configuration:")
        print(f"Search type: {retriever.search_type}")
        print(f"Search kwargs: {retriever.search_kwargs}")
        if retriever.search_kwargs.get("k") == 3:
            print("Conclusion: Retriever configured with correct k value (3).")
        else:
            print(f"Warning: Retriever's k value is {retriever.search_kwargs.get('k')}, expected 3.")
    except ValueError as ve:
        print(f"Database error during retrieval: {ve}")
    except Exception as e:
        print(f"Error during sample query with retriever: {e}")
        print("Conclusion: Retriever might not be configured correctly or database has issues.")
else:
    print("Conclusion: Failed to create the retriever.")


--- Retriever Validation ---
Conclusion: Retriever successfully created.

Retrieved 3 documents for a sample query.
Sample of retrieved document content (first 100 chars):
Symptoms and Signs
The classic symptoms of acute appendicitis are epigastric or periumbilical pain f...

Retriever configuration:
Search type: similarity
Search kwargs: {'k': 3}
Conclusion: Retriever configured with correct k value (3).


In [28]:
# System message instructing the LLM to only answer using Merck Manual 19th Edition
qna_system_message = """
You are a helpful assistant trained to answer questions based only on the Merck Manual of Medical Diagnosis and Therapy, Nineteenth Edition.
Use the context provided to find accurate and reliable answers.
If the answer is not found in the context, reply with "I don't know".
Do not mention the context or the Merck Manual in your final answer.
"""

In [29]:
# Template for formatting the user's input with context from the Merck Manual, 19th Edition and the actual medical question.
qna_user_message_template = """
###Context
The following excerpts are from the Merck Manual of Medical Diagnosis and Therapy, Nineteenth Edition:
{context}

###Question
{question}
"""

In [30]:
# --- RAG Response Function ---
def generate_rag_response1(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks1 = retriever.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks1]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

In [31]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
print(generate_rag_response1(user_input1))

/tmp/ipython-input-2450110612.py:5: LangChainDeprecationWarning: The method `BaseRetriever.get_relevant_documents` was deprecated in langchain-core 0.1.46 and will be removed in 1.0. Use :meth:`~invoke` instead.
  relevant_document_chunks1 = retriever.get_relevant_documents(query=user_input,k=k)
Llama.generate: 1 prefix-match hit, remaining 466 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   36526.28 ms /   466 tokens (   78.38 ms per token,    12.76 tokens per second)
llama_perf_context_print:        eval time =   18035.72 ms /   114 runs   (  158.21 ms per token,     6.32 tokens per second)
llama_perf_context_print:       total time =   54624.20 ms /   580 tokens
llama_perf_context_print:    graphs reused =        109


###Answer
The protocol for managing sepsis in a critical care unit includes keeping the patient warm, controlling hemorrhage, checking and providing respiratory assistance if necessary, avoiding food or fluids by mouth, turning the patient's head to prevent aspiration if emesis occurs, draining abscesses, excising necrotic tissues, administering antibiotics, and normalizing blood glucose levels. This should be continued for at least 5 days after shock resolves and evidence of infection subsides.


In [32]:
#Fine tuning rag - remove chunk overlap
# Initializing a RecursiveCharacterTextSplitter to split the text into manageable chunks for embedding and retrieval
text_splitter1 = RecursiveCharacterTextSplitter.from_tiktoken_encoder(
    encoding_name='cl100k_base',
    chunk_size=512,
    chunk_overlap= 0
)

In [33]:
document_chunks1 = loader.load_and_split(text_splitter1)

In [34]:
#Checking the number of text chunks the pdf has been split into
len(document_chunks1)

8461

In [35]:
# Generating embedding for the first document chunk
embedding_1 = embedding_function.embed_query(document_chunks1[0].page_content)
# Generating embedding for the second document chunk
embedding_2 = embedding_function.embed_query(document_chunks1[1].page_content)

In [36]:
vectorstore1 = Chroma.from_documents( #creating a Chroma vector store from a set of document chunks.
    document_chunks1, #creating a list of text chunks that will be converted into embeddings..
    embedding_function, #model responsible for embedding the document chunks into vector representations
    persist_directory=persist_directory #name of the collection in the Chroma database
)

In [37]:
 #Loading Chroma vector store with the given embedding model
vectorstore1 = Chroma(persist_directory=persist_directory,embedding_function=embedding_function)

/tmp/ipython-input-3113433777.py:2: LangChainDeprecationWarning: The class `Chroma` was deprecated in LangChain 0.2.9 and will be removed in 1.0. An updated version of the class exists in the :class:`~langchain-chroma package and should be used instead. To use it run `pip install -U :class:`~langchain-chroma` and import as `from :class:`~langchain_chroma import Chroma``.
  vectorstore1 = Chroma(persist_directory=persist_directory,embedding_function=embedding_function)


In [38]:
#Accessing the embedding function used in the Chroma vector store
vectorstore1.embeddings

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False, 'architecture': 'BertModel'})
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
), model_name='all-MiniLM-L6-v2', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False, show_progress=False)

In [39]:
retriever1 = vectorstore1.as_retriever( #Converting the Chroma vector store into a retriever for querying.
    search_type='similarity', #Specifying that retrieval is based on cosine similarity
    search_kwargs={'k': 3} #Retrieving the top 3 most similar documents for a given query.
)

In [40]:
def generate_rag_response2(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks1 = retriever1.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks1]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

In [41]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
print(generate_rag_response2(user_input1))

Llama.generate: 329 prefix-match hit, remaining 592 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   45967.45 ms /   592 tokens (   77.65 ms per token,    12.88 tokens per second)
llama_perf_context_print:        eval time =   20594.92 ms /   127 runs   (  162.16 ms per token,     6.17 tokens per second)
llama_perf_context_print:       total time =   66632.63 ms /   719 tokens
llama_perf_context_print:    graphs reused =        122


Based on the context provided, the protocol for managing sepsis in a critical care unit includes:
1. Controlling hemorrhage.
2. Checking and assisting with airway and ventilation.
3. Providing adequate nutrition.
4. Preventing infection, stress ulcers and gastritis, and pulmonary embolism.
5. Monitoring vital signs, fluid intake and output, and weight.
6. Conducting routine daily blood tests, including electrolytes and a CBC, and additional tests as needed.
7. Providing ongoing and


Observations:
With k=3, is slightly more detailed and includes examples of cases (like when surgery is impossible).

With k=2, is slightly shorter and cuts off earlier, suggesting that retrieving more chunks (k=3) provided additional helpful context for a more complete answer.

In [42]:
user_input2 = 'What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'
print(generate_rag_response1(user_input2))

Llama.generate: 116 prefix-match hit, remaining 554 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   43167.11 ms /   554 tokens (   77.92 ms per token,    12.83 tokens per second)
llama_perf_context_print:        eval time =   15976.42 ms /   101 runs   (  158.18 ms per token,     6.32 tokens per second)
llama_perf_context_print:       total time =   59195.98 ms /   655 tokens
llama_perf_context_print:    graphs reused =         96


###Answer
The common symptoms for appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia, which shifts to the right lower quadrant after a few hours. The pain increases with cough and motion. Direct and rebound tenderness can be found at McBurney's point. Appendicitis is typically treated with surgical removal. There is no cure for appendicitis with medicine alone.


In [43]:
#Fine tuning with top 2 answers
retriever2 = vector_db.as_retriever( #Converting the Chroma vector store into a retriever for querying.
    search_type='similarity', #Specifying that retrieval is based on cosine similarity
    search_kwargs={'k': 2} #Retrieving the top 2 most similar documents for a given query.
)

In [44]:
def generate_rag_response3(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever2.get_relevant_documents(query=user_input,k=k)
    context_list = [d.page_content for d in relevant_document_chunks]

    # Combine document chunks into a single context
    context_for_query = ". ".join(context_list)

    user_message = qna_user_message_template.replace('{context}', context_for_query)
    user_message = user_message.replace('{question}', user_input)

    prompt = qna_system_message + '\n' + user_message

    # Generate the response
    try:
        response = llm(
                  prompt=prompt,
                  max_tokens=max_tokens,
                  temperature=temperature,
                  top_p=top_p,
                  top_k=top_k
                  )

        # Extract and print the model's response
        response = response['choices'][0]['text'].strip()
    except Exception as e:
        response = f'Sorry, I encountered the following error: \n {e}'

    return response

In [45]:
user_input2 = 'What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'
print(generate_rag_response3(user_input2))

Llama.generate: 669 prefix-match hit, remaining 1 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =       0.00 ms /     1 tokens (    0.00 ms per token,      inf tokens per second)
llama_perf_context_print:        eval time =   16169.98 ms /   102 runs   (  158.53 ms per token,     6.31 tokens per second)
llama_perf_context_print:       total time =   16222.32 ms /   103 tokens
llama_perf_context_print:    graphs reused =         97


###Answer
The common symptoms for appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia, which shifts to the right lower quadrant after a few hours. The pain increases with cough and motion. Direct and rebound tenderness can be found at McBurney's point. Appendicitis is typically treated with surgical removal. There is no cure for appendicitis with medicine alone.


Observations on question2:

Removing chunk overlap did not change the answers in this case, probably because the relavant information for the question was already contained within a single chunk.

Removing chunk overlap might affect the answer if the question required information that spanned across multiple chunks from different parts of the text.


In [46]:
#Question3: What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?
user_input3 = "What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?"
print(generate_rag_response1(user_input3))

Llama.generate: 116 prefix-match hit, remaining 514 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   40422.46 ms /   514 tokens (   78.64 ms per token,    12.72 tokens per second)
llama_perf_context_print:        eval time =   20201.26 ms /   127 runs   (  159.07 ms per token,     6.29 tokens per second)
llama_perf_context_print:       total time =   60694.20 ms /   641 tokens
llama_perf_context_print:    graphs reused =        122


###Answer
Sudden patchy hair loss, also known as alopecia areata, can be treated with various options including topical corticosteroids, intralesional corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The underlying cause should be identified and treated if it's not alopecia areata. Possible causes include androgenetic alopecia, drugs


In [47]:
#Fine Tuning - LLM Parameters
#Temperature = 0.7
user_input3 = 'What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'
print(generate_rag_response1(user_input3, temperature=0.7))

Llama.generate: 629 prefix-match hit, remaining 1 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =       0.00 ms /     1 tokens (    0.00 ms per token,      inf tokens per second)
llama_perf_context_print:        eval time =   20427.58 ms /   128 runs   (  159.59 ms per token,     6.27 tokens per second)
llama_perf_context_print:       total time =   20499.61 ms /   129 tokens
llama_perf_context_print:    graphs reused =        123


###Answer
There are several treatment options for addressing sudden patchy hair loss, also known as alopecia areata. These include topical corticosteroids, intralesional corticosteroids, topical minoxidil, topical anthralin, topical immunotherapy (diphencyprone or squaric acid dibutylester), or psoralen plus ultraviolet A (PUVA). The choice of treatment depends on the severity and extent of hair loss. Underlying disorders, such as virilization in women or scarring hair loss, should


Observations:
With temperature = 0, is more precise and structured.

With temperature = 0.7, introduces a bit more variability, with slight changes in wording, reflecting the increased randomness from the higher temperature.

In [48]:
user_input4 = 'What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'
print(generate_rag_response1(user_input4))

Llama.generate: 116 prefix-match hit, remaining 374 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   28816.69 ms /   374 tokens (   77.05 ms per token,    12.98 tokens per second)
llama_perf_context_print:        eval time =   20142.06 ms /   127 runs   (  158.60 ms per token,     6.31 tokens per second)
llama_perf_context_print:       total time =   49029.65 ms /   501 tokens
llama_perf_context_print:    graphs reused =        122


###Answer
The recommended treatments for a person with a brain injury include physical and occupational therapy, skill-building activities, counseling, and participation in brain injury support groups. For those with prolonged comas, a prolonged period of rehabilitation, particularly in cognitive and emotional areas, may be required. There is no specific medical treatment for brain injuries, and drugs that slow the symptomatic progression of dementia do not appear beneficial. Patients should be reevaluated regularly, and treatment priorities are based on the severity of cognitive dysfunction. Extensive cognitive therapy is often required for those with


In [49]:
#Fine Tuning - LLM Parameters
#Top_p = 0.8
user_input4 = 'What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'
print(generate_rag_response1(user_input4, top_p=0.8))

Llama.generate: 489 prefix-match hit, remaining 1 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =       0.00 ms /     1 tokens (    0.00 ms per token,      inf tokens per second)
llama_perf_context_print:        eval time =   20189.63 ms /   128 runs   (  157.73 ms per token,     6.34 tokens per second)
llama_perf_context_print:       total time =   20260.70 ms /   129 tokens
llama_perf_context_print:    graphs reused =        123


###Answer
The recommended treatments for a person with a brain injury include a team approach that combines physical, occupational, and speech therapy, skill-building activities, and counseling to meet the patient's social and emotional needs. Brain injury support groups may also provide assistance to the families of brain-injured patients. For patients whose coma exceeds 24 hours and have major persistent neurologic sequelae, a prolonged period of rehabilitation, particularly in cognitive and emotional areas, is often required. Physical and occupational therapy may improve functioning and make the environment safer for the patient


Observations:
With Top_p = 0.95 means the model will consider the top 95% probability mass of the next token and select from that pool. This makes the answer generated more focused, coherent responses because it prioritizes more likely words and phrases.

With Top_p = 0.8 means the model restricts the token selection to only the top 80% of the probability distribution, including less likely but more diverse options.

The model is focusing more on the main facts with 0.95 temp results(like supportive care and rehab) without adding unnecessary or unrelated details. In .80 top_p tuning, the model explores a wider range of possibilities, leading to additional information about severe cases, surgery, and treatment phases.

In [50]:
user_input5 = 'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'
print(generate_rag_response1(user_input5))

Llama.generate: 116 prefix-match hit, remaining 913 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   71027.48 ms /   913 tokens (   77.80 ms per token,    12.85 tokens per second)
llama_perf_context_print:        eval time =   20850.23 ms /   127 runs   (  164.18 ms per token,     6.09 tokens per second)
llama_perf_context_print:       total time =   91948.98 ms /  1040 tokens
llama_perf_context_print:    graphs reused =        122


Based on the context provided, the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip include:
1. Rest: The person should avoid putting weight on the injured leg to prevent further damage.
2. Ice and compression: Applying ice and compression with the use of a thigh sleeve can help reduce swelling and pain.
3. Pain management: Nonsteroidal anti-inflammatory drugs (NSAIDs) and analgesics may be prescribed as necessary.
4. Immobilization: A splint or cast may be required to keep


In [51]:
#Fine Tuning - LLM Parameters
#Top_k = 25
user_input5 = 'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'
print(generate_rag_response1(user_input5, top_k=25))

Llama.generate: 1028 prefix-match hit, remaining 1 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =       0.00 ms /     1 tokens (    0.00 ms per token,      inf tokens per second)
llama_perf_context_print:        eval time =   20863.25 ms /   128 runs   (  162.99 ms per token,     6.14 tokens per second)
llama_perf_context_print:       total time =   20932.99 ms /   129 tokens
llama_perf_context_print:    graphs reused =        123


Based on the context provided, the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip include:
1. Rest: The person should avoid putting weight on the injured leg to prevent further damage.
2. Ice and compression: Applying ice and compression with the use of a thigh sleeve can help reduce swelling and pain.
3. Pain management: Nonsteroidal anti-inflammatory drugs (NSAIDs) and analgesics may be prescribed as necessary.
4. Immobilization: A splint or cast may be required to keep


Observations:
In this case, the generated answers are the same, probably because the topic is very structured, and the model chooses almost the same tokens even when it has a broader range of options (top_k=50).

In [52]:
#Output Evaluation
groundedness_rater_system_message = """

You will be presented a ###Question, ###Context used by the AI system and AI generated ###Answer.

Your task is to judge the extent to which the ###Answer is derived from ###Context.

Rate it 1 - if The ###Answer is not derived from the ###Context at all
Rate it 2 - if The ###Answer is derived from the ###Context only to a limited extent
Rate it 3 - if The ###Answer is derived from ###Context to a good extent
Rate it 4 - if The ###Answer is derived from ###Context mostly
Rate it 5 - if The ###Answer is is derived from ###Context completely

Please note: Make sure you give a single overall rating in the range of 1 to 5 along with an overall explanation.

"""

In [53]:
relevance_rater_system_message = """

You will be presented with a ###Question, the ###Context used by the AI system to generate a response, and the AI-generated ###Answer.

Your task is to judge the extent to which the ###Answer is relevant to the ###Question, considering whether it directly addresses the key aspects of the ###Question based on the provided ###Context.

Rate the relevance as follows:
- Rate 1 – The ###Answer is not relevant to the ###Question at all.
- Rate 2 – The ###Answer is only slightly relevant to the **###Question**, missing key aspects.
- Rate 3 – The ###Answer is moderately relevant, addressing some parts of the **###Question** but leaving out important details.
- Rate 4 – The ###Answer is mostly relevant, covering key aspects but with minor gaps.
- Rate 5 – The ###Answer is fully relevant, directly answering all important aspects of the **###Question** with appropriate details from the **###Context**.

Note: Provide a single overall rating in the range of 1 to 5, along with a brief explanation of why you assigned that score.
"""

In [54]:
user_message_template = """
###Question
{question}

###Context
{context}

###Answer
{answer}
"""

In [55]:
#Function for output evaluation
def generate_ground_relevance_response(user_input,k=3,max_tokens=128,temperature=0,top_p=0.95,top_k=50):
    global qna_system_message,qna_user_message_template
    # Retrieve relevant document chunks
    relevant_document_chunks = retriever.get_relevant_documents(query=user_input,k=3)
    context_list = [d.page_content for d in relevant_document_chunks]
    context_for_query = ". ".join(context_list)

    # Combine user_prompt and system_message to create the prompt
    prompt = f"""[INST]{qna_system_message}\n
                {'user'}: {qna_user_message_template.format(context=context_for_query, question=user_input)}
                [/INST]"""

    response = llm(
            prompt=prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            echo=False
            )

    answer =  response["choices"][0]["text"]

    # Combine user_prompt and system_message to create the prompt
    groundedness_prompt = f"""[INST]{groundedness_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    # Combine user_prompt and system_message to create the prompt
    relevance_prompt = f"""[INST]{relevance_rater_system_message}\n
                {'user'}: {user_message_template.format(context=context_for_query, question=user_input, answer=answer)}
                [/INST]"""

    response_1 = llm(
            prompt=groundedness_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            echo=False
            )

    response_2 = llm(
            prompt=relevance_prompt,
            max_tokens=max_tokens,
            temperature=temperature,
            top_p=top_p,
            top_k=top_k,
            stop=['INST'],
            echo=False
            )

    return response_1['choices'][0]['text'],response_2['choices'][0]['text']

In [56]:
user_input1 = 'What is the protocol for managing sepsis in a critical care unit?'
ground,rel = generate_ground_relevance_response(user_input1,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: 1 prefix-match hit, remaining 935 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   73768.81 ms /   935 tokens (   78.90 ms per token,    12.67 tokens per second)
llama_perf_context_print:        eval time =   21300.98 ms /   132 runs   (  161.37 ms per token,     6.20 tokens per second)
llama_perf_context_print:       total time =   95142.85 ms /  1067 tokens
llama_perf_context_print:    graphs reused =        127
Llama.generate: 5 prefix-match hit, remaining 1133 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   89201.44 ms /  1133 tokens (   78.73 ms per token,    12.70 tokens per second)
llama_perf_context_print:        eval time =   23680.94 ms /   144 runs   (  164.45 ms per token,     6.08 tokens per second)
llama_perf_context_print:       total time =  112965.03 ms /  1277 tokens
llama_perf_context_print:    

 Rating: 5 - The Answer is derived from the Context completely.

Explanation: The Context provides detailed information about the management of critically ill patients in an ICU, specifically mentioning sepsis as a common condition and the importance of prompt and adequate therapy. The Answer summarizes the key points from the Context regarding the protocol for managing sepsis in a critical care unit, including controlling hemorrhage, checking and providing respiratory assistance, keeping the patient warm, providing adequate nutrition and prevention of complications, monitoring vital signs and using complex devices, and performing routine and additional blood tests. Therefore, the Answer is a comprehensive and accurate reflection of the information provided in the Context.

 Rating: 5

Explanation: The AI-generated answer directly addresses all the key aspects of the question based on the provided context. It mentions controlling hemorrhage, checking and providing respiratory assistanc

Observations

The rating of 5 for groundness and relavance indicates that the answer is fully grounded in the context and is also highly relevant to the query.

In [57]:
user_input2 = 'What are the common symptoms for appendicitis, and can it be cured via medicine? If not, what surgical procedure should be followed to treat it?'
ground,rel = generate_ground_relevance_response(user_input2,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: 5 prefix-match hit, remaining 680 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   53093.59 ms /   680 tokens (   78.08 ms per token,    12.81 tokens per second)
llama_perf_context_print:        eval time =   16253.50 ms /   101 runs   (  160.93 ms per token,     6.21 tokens per second)
llama_perf_context_print:       total time =   69401.02 ms /   781 tokens
llama_perf_context_print:    graphs reused =         97
Llama.generate: 5 prefix-match hit, remaining 851 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   66677.44 ms /   851 tokens (   78.35 ms per token,    12.76 tokens per second)
llama_perf_context_print:        eval time =   23952.96 ms /   147 runs   (  162.95 ms per token,     6.14 tokens per second)
llama_perf_context_print:       total time =   90714.60 ms /   998 tokens
llama_perf_context_print:    g

 Rating: 5

Explanation: The AI answer is completely derived from the context. The context clearly states that the symptoms of appendicitis include epigastric or periumbilical pain followed by brief nausea, vomiting, and anorexia, which shifts to the right lower quadrant after a few hours. The context also states that the pain increases with cough and motion, and other symptoms may include fever, loss of appetite, and constipation or diarrhea. The context also states that appendicitis cannot be cured with medicine alone, and surgical removal of the appendix is required to treat the condition. The AI answer accurately reflects this information from the context.

 Rating: 5

Explanation: The AI-generated answer directly addresses all the important aspects of the question. It correctly identifies the common symptoms of appendicitis, which includes epigastric or periumbilical pain, nausea, vomiting, anorexia, and right lower quadrant pain that increases with cough and motion. It also menti

In [58]:
user_input3 = 'What are the effective treatments or solutions for addressing sudden patchy hair loss, commonly seen as localized bald spots on the scalp, and what could be the possible causes behind it?'
ground,rel = generate_ground_relevance_response(user_input3,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: 5 prefix-match hit, remaining 640 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   49845.99 ms /   640 tokens (   77.88 ms per token,    12.84 tokens per second)
llama_perf_context_print:        eval time =   20939.43 ms /   130 runs   (  161.07 ms per token,     6.21 tokens per second)
llama_perf_context_print:       total time =   70859.69 ms /   770 tokens
llama_perf_context_print:    graphs reused =        125
Llama.generate: 5 prefix-match hit, remaining 840 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   64617.15 ms /   840 tokens (   76.93 ms per token,    13.00 tokens per second)
llama_perf_context_print:        eval time =   12832.54 ms /    79 runs   (  162.44 ms per token,     6.16 tokens per second)
llama_perf_context_print:       total time =   77489.57 ms /   919 tokens
llama_perf_context_print:    g

 Rating: 5

Explanation: The AI-generated answer is completely derived from the context. The context provides detailed information about the causes and treatments for alopecia areata, and the answer accurately summarizes that information. The answer also correctly identifies possible causes of sudden patchy hair loss beyond alopecia areata, which is also mentioned in the context.

 Rating: 5

Explanation: The AI-generated answer directly addresses the key aspects of the question by providing a list of effective treatments for sudden patchy hair loss (alopecia areata) and mentioning the possible causes, which include androgenetic alopecia, drugs, and infection. The context provided in the text also discusses these same treatments and causes for alopecia areata.


Observations

The rating of 5 for groundness and relavance indicates that the answer is fully grounded in the context and is also highly relevant to the query.

In [59]:
user_input4 = 'What treatments are recommended for a person who has sustained a physical injury to brain tissue, resulting in temporary or permanent impairment of brain function?'
ground,rel = generate_ground_relevance_response(user_input4,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: 5 prefix-match hit, remaining 500 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   37660.68 ms /   500 tokens (   75.32 ms per token,    13.28 tokens per second)
llama_perf_context_print:        eval time =   24943.07 ms /   156 runs   (  159.89 ms per token,     6.25 tokens per second)
llama_perf_context_print:       total time =   62696.70 ms /   656 tokens
llama_perf_context_print:    graphs reused =        150
Llama.generate: 5 prefix-match hit, remaining 726 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   56961.87 ms /   726 tokens (   78.46 ms per token,    12.75 tokens per second)
llama_perf_context_print:        eval time =   21261.38 ms /   131 runs   (  162.30 ms per token,     6.16 tokens per second)
llama_perf_context_print:       total time =   78297.40 ms /   857 tokens
llama_perf_context_print:    g

 Rating: 5 - The AI-generated answer is derived from the context completely. The context mentions physical, occupational, and speech therapy, skill-building activities, counseling, and rehabilitation for brain injury patients. The AI-generated answer accurately reflects these points and adds some additional details, such as the importance of reevaluation and the fact that there is no specific medical treatment for brain injury. The answer also mentions cognitive therapy for patients with severe cognitive dysfunction, which is mentioned in the context but not explicitly stated as a form of therapy. Overall, the AI-generated answer accurately summarizes the information in the context.

 ###Rating: 5

###Explanation:

The AI-generated answer directly addresses the key aspects of the question by discussing the recommended treatments for a person with a physical injury to brain tissue, which includes physical, occupational, and speech therapy, skill-building activities, counseling, and cogn

In [60]:
user_input5 = 'What are the necessary precautions and treatment steps for a person who has fractured their leg during a hiking trip, and what should be considered for their care and recovery?'
ground,rel = generate_ground_relevance_response(user_input5,max_tokens=350)

print(ground,end="\n\n")
print(rel)

Llama.generate: 5 prefix-match hit, remaining 1039 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   81378.04 ms /  1039 tokens (   78.32 ms per token,    12.77 tokens per second)
llama_perf_context_print:        eval time =   21075.97 ms /   128 runs   (  164.66 ms per token,     6.07 tokens per second)
llama_perf_context_print:       total time =  102525.67 ms /  1167 tokens
llama_perf_context_print:    graphs reused =        123
Llama.generate: 5 prefix-match hit, remaining 1236 prompt tokens to eval
llama_perf_context_print:        load time =    1192.15 ms
llama_perf_context_print: prompt eval time =   97816.99 ms /  1236 tokens (   79.14 ms per token,    12.64 tokens per second)
llama_perf_context_print:        eval time =   17463.16 ms /   104 runs   (  167.91 ms per token,     5.96 tokens per second)
llama_perf_context_print:       total time =  115335.38 ms /  1340 tokens
llama_perf_context_print:   

 Rating: 5

Explanation: The AI-generated answer is completely derived from the context. The context provides detailed information about the necessary precautions and treatment steps for a person with a fractured leg, including RICE therapy, the use of crutches, splinting or casting, and the possibility of surgery in severe cases. The AI answer accurately reflects this information. Additionally, the context mentions the importance of assessing for associated injuries, which is also included in the AI answer.

 Rating: 5

Explanation: The AI-generated answer directly addresses all the key aspects of the question, including necessary precautions (RICE, splinting or casting, crutches), treatment steps (NSAIDs and analgesics, surgery), and considerations for care and recovery (rehabilitation with stretching and strengthening exercises, assessment for associated injuries). The answer also includes relevant details from the context, such as the importance of RICE and the potential need for s

Observations

The rating of 5 for groudness and relavance indicates that the answer is fully grounded in the context and is also highly relevant to the query.

Actionable Insights and Business Recommendations
Business Insights

Accuracy and Trustworthiness

RAG-based answers are context-specific, reliable, and trustworthy, enhancing customer satisfaction in medical applications.
Enhanced User Experience

The RAG model is able to provide personalized, relevant responses and more accurate answers.
Cost Reduction in Medical Consultations

The model can be used for automating medical Q&A, reducing the need for consultations, saving costs in telemedicine and healthcare support.
Recommendations

Continuous Fine-tuning

The model should be updated and fine-tuned to reflect the latest medical guidelines and research.
Ethical Considerations

Data privacy should be prioritized to ensure compliance with regulations like HIPAA and GDPR.
Specialized Models

Dedicated models can be created, tailored to individual medical specialties such as orthopedics, gastroenterology, oncology, and more for improved accuracy and relevance.